unixfs/src/walk.rs

```diff
@@ -0,0 +1,1040 @@
+#![allow(unused, dead_code)]
+
+use crate::dir::{
+    check_directory_supported, check_hamtshard_supported, ShardError, UnexpectedDirectoryProperties,
+};
+use crate::file::visit::{Cache, FileVisit, IdleFileVisit};
+use crate::file::{FileError, FileReadFailed};
+use crate::pb::{FlatUnixFs, PBLink, PBNode, ParsingFailed, UnixFsType};
+use crate::{InvalidCidInLink, Metadata, UnexpectedNodeType};
+use cid::Cid;
+use either::Either;
+use std::borrow::Cow;
+use std::convert::TryFrom;
+use std::fmt;
+use std::path::{Path, PathBuf};
+
+/// Walker helps with walking an UnixFS tree, including all of the content and files. Created with
+/// `Walker::new` and walked over each block with `Walker::continue_block`. Use
+/// `Walker::pending_links` to learn the next [`Cid`] to be loaded and the prefetchable links.
+#[derive(Debug)]
+pub struct Walker {
+    /// This is `None` until the first block has been visited. Failing any unwraps would be logic
+    /// errors.
+    current: Option<InnerEntry>,
+    /// On the next call to `continue_walk` this will be the block, unless we have an ongoing file
+    /// walk in which case we shortcircuit to continue it. Failing any of the unwrappings of
+    /// `self.next` would be an logic error
+    next: Option<(Cid, String, usize)>,
+    pending: Vec<(Cid, String, usize)>,
+    // tried to recycle the names but that was consistently as fast and used more memory than just
+    // cloning the strings
+}
+
+/// Converts a link of a Directory, specifically not a link of HAMTShard.
+fn convert_link(
+    depth: usize,
+    nth: usize,
+    link: PBLink<'_>,
+) -> Result<(Cid, String, usize), InvalidCidInLink> {
+    let hash = link.Hash.as_deref().unwrap_or_default();
+    let cid = match Cid::try_from(hash) {
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
+    let name = match link.Name {
+        Some(Cow::Borrowed(s)) if !s.is_empty() => s.to_owned(),
+        None | Some(Cow::Borrowed(_)) => todo!("link cannot be empty"),
+        Some(Cow::Owned(s)) => s,
+    };
+    assert!(!name.contains('/'));
+    Ok((cid, name, depth))
+}
+
+/// Converts a link of HAMTShard, specifically not a link of Directory.
+fn convert_sharded_link(
+    depth: usize,
+    nth: usize,
+    link: PBLink<'_>,
+) -> Result<(Cid, String, usize), InvalidCidInLink> {
+    let hash = link.Hash.as_deref().unwrap_or_default();
+    let cid = match Cid::try_from(hash) {
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
+    let (depth, name) = match link.Name {
+        Some(Cow::Borrowed(s)) if s.len() > 2 => (depth, s[2..].to_owned()),
+        Some(Cow::Borrowed(s)) if s.len() == 2 => (depth - 1, String::from("")),
```

Comment:
shouldn't we `assert` that `depth >= 1` somewhere above? or is it ensured at the callsite? might be a good idea to comment on that

Reply:
Yeah caller calls this with `depth + 1` and now there's only a single callsite left. Getting the depth go correctly, where `depth == Path::ancestors().count()`, has been a source of frustration.

I am hesitant to add any more assertions though, as all the test cases in this file do the walking with empty name (turns into minimal depth) and a non-empty name.

Reply:
Well, I could have two parameters here describing more what's the difference. Pushing that soonish.

Reply:
Pushed; no more `depth - 1` but `nested_depth` and `sibling_depth` args, which are used as is in this link conversion.